File: web/src/types/event.ts
import type { CircleCategory } from "./circle";

export interface FriendlyEvent {
  id: string;
  title: string;
  description: string;
  circleId?: string;
  circleName?: string;
  hostId: string;
  hostName: string;
  location: EventLocation;
  startDate: Date;
  endDate: Date;
  maxAttendees: number;
  attendeeIds: string[];
  waitlistIds: string[];
  category: CircleCategory;
  imageURL?: string;
  isOutdoor: boolean;
  cost: EventCost;
  createdAt: Date;
  isCancelled: boolean;
}

export interface EventLocation {
  name: string;
  address: string;
  latitude: number;
  longitude: number;
}

export type EventCost =
  | { type: "free" }
  | { type: "fixed"; amount: number }
  | { type: "splitEvenly" };

export function formatEventCost(cost: EventCost): string {
  switch (cost.type) {
    case "free":
      return "Free";
    case "fixed":
      return `$${cost.amount.toFixed(0)}`;
    case "splitEvenly":
      return "Split evenly";
  }
}

export function formatEventDate(date: Date): string {
  return date.toLocaleDateString("en-US", {
    weekday: "short",
    month: "short",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
  });
}

export function formatEventTimeRange(start: Date, end: Date): string {
  const fmt = (d: Date) =>
    d.toLocaleTimeString("en-US", { hour: "numeric", minute: "2-digit" });
  return `${fmt(start)} – ${fmt(end)}`;
}

export function isEventUpcoming(event: FriendlyEvent): boolean {
  return event.startDate > new Date();
}

export function isEventFull(event: FriendlyEvent): boolean {
  return event.attendeeIds.length >= event.maxAttendees;
}

File: ios/Models/Event.swift
import Foundation
import FirebaseFirestore

struct Event: Identifiable, Codable, Hashable {
    @DocumentID var id: String?
    var title: String
    var description: String
    var circleId: String?
    var circleName: String?
    var hostId: String
    var hostName: String
    var location: EventLocation
    var startDate: Date
    var endDate: Date
    var maxAttendees: Int
    var attendeeIds: [String]
    var waitlistIds: [String]
    var category: CircleCategory
    var imageURL: String?
    var isOutdoor: Bool
    var cost: EventCost
    var createdAt: Date
    var isCancelled: Bool

    var attendeeCount: Int { attendeeIds.count }
    var isFull: Bool { attendeeIds.count >= maxAttendees }
    var hasWaitlist: Bool { !waitlistIds.isEmpty }
    var isUpcoming: Bool { startDate > Date() }
    var isPast: Bool { endDate < Date() }

    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEE, MMM d · h:mm a"
        return formatter.string(from: startDate)
    }

    var formattedTimeRange: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"
        return "\(formatter.string(from: startDate)) – \(formatter.string(from: endDate))"
    }

    static var placeholder: Event {
        Event(
            title: "Morning Run",
            description: "A casual 5K run through the park",
            hostId: "",
            hostName: "Alex",
            location: .init(name: "Central Park", address: "123 Park Ave", latitude: 40.785091, longitude: -73.968285),
            startDate: Date().addingTimeInterval(86400),
            endDate: Date().addingTimeInterval(86400 + 3600),
            maxAttendees: 10,
            attendeeIds: [],
            waitlistIds: [],
            category: .fitness,
            isOutdoor: true,
            cost: .free,
            createdAt: Date(),
            isCancelled: false
        )
    }
}

struct EventLocation: Codable, Hashable {
    var name: String
    var address: String
    var latitude: Double
    var longitude: Double
}

enum EventCost: Codable, Hashable {
    case free
    case fixed(amount: Double)
    case splitEvenly

    var displayText: String {
        switch self {
        case .free: return "Free"
        case .fixed(let amount): return String(format: "$%.0f", amount)
        case .splitEvenly: return "Split evenly"
        }
    }
}
